src/bigartm/cooccurrence_dictionary.cc

```diff
@@ -0,0 +1,442 @@
+  // 4. Cut every document into words, search for them in dictionary and for
+  // valid calculate co-occurrence, number of documents where these words
+  // were found close enough (in a window with width window_width) and save
+  // all in map (its own for each document)
+  // 5. merge all maps into one
+  // 6. If resulting map isn't empty create a batch and dump all information
+  // on external storage
+  // Repeat 1-6 for all portions (can work in parallel for different portions)
+
+  std::filebuf fb;
+  if (!fb.open(path_to_vw, std::ios::in)) {
+    std::cerr << "Failed to open vocab\n";
+    throw 1;
+  }
+  std::istream VowpalWabbitDoc(&fb);
+  std::mutex read_lock, write_lock;
+
+  auto func = [&]() {
+    while (true) {
+      std::vector<std::string> portion;
+
+      {
+        std::lock_guard<std::mutex> guard(read_lock);
+        if (VowpalWabbitDoc.eof())
+          return;
+
+        std::string str;
+        while (portion.size() < ITEMS_PER_BATCH) {
+          getline(VowpalWabbitDoc, str);
+          if (VowpalWabbitDoc.eof())
+            break;
+          portion.push_back(str);
+        }
+      }
+
+      if (!portion.size())
+        continue;
+
+      // First elem in external map is first_token_id, in internal it's
+      // second_token_id
+      std::map<int, std::map<int, cooccurrence_info>> cooc_map;
+
+      for (int doc_id = 0; doc_id < (int64_t) portion.size(); ++doc_id) {
+        std::vector<std::string> doc;
+        boost::split(doc, portion[doc_id], boost::is_any_of(" \t\r"));
+        if (doc.size() <= 1)
+          continue;
+
+        for (int j = 1; j < (int64_t) doc.size() - 1; ++j) {
+          auto first_token = dictionary.find(doc[j]);
+          if (first_token == dictionary.end())
+            continue;
+          int first_token_id = first_token->second;
+
+          for (int k = 1; k <= window_width && j + k < (int64_t) doc.size(); ++k) {
+            auto second_token = dictionary.find(doc[j + k]);
+            if (second_token == dictionary.end())
+              continue;
+            int second_token_id = second_token->second;
+            if (first_token_id == second_token_id)
+              continue;
+
+            SavePairOfTokens(first_token_id, second_token_id, doc_id, cooc_map);
+
+            SavePairOfTokens(second_token_id, first_token_id, doc_id, cooc_map);
+          }
+        }
+      }
+
+      {
+        std::lock_guard<std::mutex> guard(write_lock);
+        if (!cooc_map.empty())
+          UploadBatchOnDisk(batch_manager, cooc_map);
+      }
+    }
+  };
+  int num_of_threads = std::thread::hardware_concurrency();
+  if (!num_of_threads)
+    num_of_threads = 1;
+  std::vector<std::shared_future<void>> tasks;
+  for (int i = 0; i < num_of_threads; ++i)
+    tasks.push_back(std::move(std::async(std::launch::async, func)));
+  for (int i = 0; i < num_of_threads; ++i)
+    tasks[i].get();
+}
+
+inline void CooccurrenceDictionary::ReadAndMergeBatches() {
```

Comment:
There's no need in inline somewhere in your changes.
Inline has two functions
1) ask compiler to put you function code into the place of call without real call for speed-up (it is really obsolete now, because nowadays compilers can make such decisions by themselves)

2) declaring functions in .h file:
 FYI if you define this function in .h file
> int foo() { return 5; }
and then include it into two .cpp files, you'll get error "Multiple definitions of foo()". To avoid it you can write
> **inline** int foo() { return 5; }

But in this review you have no cases of usage of inline, so remove all it's entries.